### DISCUSSION TITLE
More than one table in output

### DISCUSSION THREAD

**Discussion by kubusek17** [dbosoft/YaNco#266] [Q&A] [answered]
Hello, is it possible to process more than one table in output ?

**Comment by fw2568** [comment]
Sure, you can call GetField, MapStructure and MapTable as often as you need. However, you are probably struggling with the fact that you can only return one **type**. 

The concept is, that you can build your output type within the output method. Example:

https://github.com/dbosoft/sap-dmsclient/blob/blog_series/part-2/src/SAPDms.Core/Functions/ReadDocument.cs

In this example, we call both MapStructure and MapTable and create the output type from their results. Of course, you can do this for multiple tables.

**Reply by kubusek17** [reply]
It works only with a structure and a table, with two tables I have a problem

Following code 

```c#
from _ in func.HandleReturn()
from docData in func.MapTable("HEADERS", docData =>
    from status in docData.GetField<string>("VBELN")
    from description in docData.GetField<string>("FKART")
    select new {status, description })
from fileData in func.MapTable("CUSTOMERS", s =>
        from originalType in s.GetField<string>("KUNNR")
        from filePath in s.GetField<string>("NAME1")
        from applicationId in s.GetField<string>("NAME2")
        from fileId in s.GetField<string>("ORT01")
        select new DocumentFileInfo(
            applicationId, originalType, fileId , Path.GetFileName(filePath))
    )
select new DocumentData(docData.description, docData.status, fileData.ToArr())
``` 

generates errors I can not understand 

Error	CS0029	Cannot implicitly convert type 'LanguageExt.Either<Dbosoft.YaNco.RfcErrorInfo, System.Collections.Generic.IEnumerable<Dbosoft.SAPDms.DocumentFileInfo>>' to 'LanguageExt.EitherRight<Dbosoft.YaNco.RfcErrorInfo>'

Error	CS8016	Transparent identifier member access failed for field 'docData' of 'LanguageExt.DataTypes.Serialisation.EitherData<Dbosoft.YaNco.RfcErrorInfo, <anonymous type: Dbosoft.YaNco.IFunction _, System.Collections.Generic.IEnumerable<<anonymous type: string status, string description>> docData>>'.  Does the data being queried implement the query pattern?	

Error	CS1061	'IEnumerable<<anonymous type: string status, string description>>' does not contain a definition for 'description' and no accessible extension method 'description' accepting a first argument of type 'IEnumerable<<anonymous type: string status, string description>>' could be found (are you missing a using directive or an assembly reference?)


Error	CS8016	Transparent identifier member access failed for field 'docData' of 'LanguageExt.DataTypes.Serialisation.EitherData<Dbosoft.YaNco.RfcErrorInfo, <anonymous type: Dbosoft.YaNco.IFunction _, System.Collections.Generic.IEnumerable<<anonymous type: string status, string description>> docData>>'.  Does the data being queried implement the query pattern?

Error	CS1061	'IEnumerable<<anonymous type: string status, string description>>' does not contain a definition for 'status' and no accessible extension method 'status' accepting a first argument of type 'IEnumerable<<anonymous type: string status, string description>>' could be found (are you missing a using directive or an assembly reference?)

**Reply by fw2568** [reply, marked as answer]
Transparent identifier members are automatic fields generated in linq statements (see https://ericlippert.com/2014/07/31/transparent-identifiers-part-one/). As your field docData is now a IEnumerable<> it contains no fields description and status. 

A working code:

``` csharp
from _ in func.HandleReturn()
from headers in func.MapTable("HEADERS", docData =>
        from documentNo in docData.GetField<string>("VBELN")
        from documentType in docData.GetField<string>("FKART")
        select new {documentNo, documentType } 
    )
from customers in func.MapTable("CUSTOMERS", s =>
        from kunnr in s.GetField<string>("KUNNR")
        from name1 in s.GetField<string>("NAME1")
        from name2 in s.GetField<string>("NAME2")
        from location in s.GetField<string>("ORT01")
        select new { kunnr, name1, name2, location }
    )
select new { headers, customers }
```

This code would give a IEnumerable headers containing the header data and a IEnumerable customers containing the customer data.

At this point I also would like to add that we offer here at dbosoft a developer subscription for YaNco that covers consulting support for such questions. So how to use linq and YaNco best practices and discount on trainings.
See also here: https://support.dbosoft.eu/hc/en-us/articles/360021454080-YaNco-Support-Plans 

Best Regards,
Frank
